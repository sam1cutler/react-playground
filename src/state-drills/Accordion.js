import React from 'react';

class Accordion extends React.Component {
    
    static defaultProps = { sections: [] }
    
    state = {
        currentSectionIndex: null
    }

    handleButtonClick = (index) => {
        this.setState( {currentSectionIndex: index } )
    }
    
    renderContent = (index) => {
        if (index === this.state.currentSectionIndex) {
            return this.props.sections[index].content
        } 
    }

    renderList() {
        return this.props.sections.map( (item, index) => (
            <li 
                key={index} 
                onClick={ () => this.handleButtonClick(index) }
            >
                <button>{item.title}</button>
                <p>{this.renderContent(index)}</p>
            </li>
        ))
    }
    
    render() {
        
        return (
            <div>
                <h2>Accordion!</h2>
                <ul>
                    {this.renderList()}
                </ul>
            </div>
        )
    }
}

export default Accordion;