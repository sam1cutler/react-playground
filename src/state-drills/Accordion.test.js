import React from 'react';
import ReactDOM from 'react-dom';
import Accordion from './Accordion';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

describe(`Accordion component`, () => {

    const accordionSections = [
        {
          title: 'Section 1',
          content: 'Lorem ipsum dolor sit amet consectetur adipisicing elit.',
        },
        {
          title: 'Section 2',
          content: 'Cupiditate tenetur aliquam necessitatibus id distinctio quas nihil ipsam nisi modi!',
        },
        {
          title: 'Section 3',
          content: 'Animi amet cumque sint cupiditate officia ab voluptatibus libero optio et?',
        },
    ]
    
    it('renders without errors', () => {
        const div = document.createElement('div');
        ReactDOM.render(<Accordion />, div);
        ReactDOM.unmountComponentAtNode(div);
    })

    it('renders an empty li when no sections prop is supplied', () => {
        const wrapper = shallow(<Accordion />)
        expect(toJson(wrapper)).toMatchSnapshot()
    })

    it('opens a clicked section', () => {
        const wrapper = shallow(<Accordion sections={accordionSections} />)
        wrapper.find('button').at(2).simulate('click')
        expect(toJson(wrapper)).toMatchSnapshot()
    })

    it('opens Section 1 after opening Section 2', () => {
        const wrapper = shallow(<Accordion sections={accordionSections} />)
        wrapper.find('button').at(1).simulate('click')
        wrapper.find('button').at(0).simulate('click')
        expect(toJson(wrapper)).toMatchSnapshot()
    })

})