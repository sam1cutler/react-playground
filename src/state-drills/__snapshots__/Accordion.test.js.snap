// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion component opens Section 1 after opening Section 2 1`] = `
<div>
  <h2>
    Accordion!
  </h2>
  <ul>
    <li
      key="0"
      onClick={[Function]}
    >
      <button>
        Section 1
      </button>
      <p />
    </li>
    <li
      key="1"
      onClick={[Function]}
    >
      <button>
        Section 2
      </button>
      <p />
    </li>
    <li
      key="2"
      onClick={[Function]}
    >
      <button>
        Section 3
      </button>
      <p />
    </li>
  </ul>
</div>
`;

exports[`Accordion component opens a clicked section 1`] = `
<div>
  <h2>
    Accordion!
  </h2>
  <ul>
    <li
      key="0"
      onClick={[Function]}
    >
      <button>
        Section 1
      </button>
      <p />
    </li>
    <li
      key="1"
      onClick={[Function]}
    >
      <button>
        Section 2
      </button>
      <p />
    </li>
    <li
      key="2"
      onClick={[Function]}
    >
      <button>
        Section 3
      </button>
      <p />
    </li>
  </ul>
</div>
`;

exports[`Accordion component renders an empty li when no sections prop is supplied 1`] = `
<div>
  <h2>
    Accordion!
  </h2>
  <ul />
</div>
`;
